src/v3/src/hooks/useOnSubmit.ts

```diff
@@ -139,6 +141,13 @@ export const useOnSubmit = (): (options: OnSubmitHandlerOptions) => Promise<void
     // Allow username transformation if applicable
     if ('identifier' in payload) {
       payload.identifier = transformIdentifier(widgetProps, step, payload.identifier as string);
+
+      // Widget rememberMe feature stores the entered identifier in a cookie, to pre-fill the form on subsequent visits to page
+      if (features?.rememberMe) {
+        setUsernameCookie(payload.identifier as string);
+      } else {
+        removeUsernameCookie();
+      }
```

Comment:
This code looks sound, but wondering if the placement is what we want. We already have very similar code in https://github.com/okta/okta-signin-widget/blob/5e12b8f1430d6dbfa16391199a71e57fe80ebdc7/src/v3/src/transformer/terminal/transformTerminalTransaction.ts#L139 but it only runs on a success terminal transaction. I'm wondering if we should not set this cookie until we reach a success state because otherwise we will overwrite the cookie before the user's authentication is successful.

Reply:
Posting what Jared said about the motivation for the original fix he made in gen2 from our DM: 

_"The rememberMyUsername feature was implemented differently in g1 and g2. g1 it stored the username in a cookie after the user typed, where g2 wait until a terminal success state was reached. I believe a customer reported an issue with MFA, where the username was being extracted from the IDX response and when MFA was enabled, the IDX pipeline lost the context of the username and feature broke. The fix was to store the username when the user entered their username and proceeded, mimicking the g1 implementation and fixed the MFA issue."_

So basically the current state of gen3 matches the state of gen2 BEFORE Jared made the fix because it only store the username cookie upon reaching a success terminal state. I was able to repro the original bug in the current state of gen3 and verify that this change does fix it. Since this cookie is only used to pre-populate the username field on the identify screen, I believe the thought is that it's safe to overwrite it whenever the user proceeds from the identify screen even if they don't reach a success terminal state because we presume that the user entered their username correctly and even if their transaction failed they would want to try again with the same username pre-populated.

Let me know if any of that isn't clear

Reply:
Quick question on this, does that mean we no longer need the logic in the terminal transaction transformer?

Reply:
That I'm not sure about, I can try removing it and see how it affects our tests?

Reply:
Does this logic still exist in Gen 2? Maybe Jared may know, I would assume he would have removed it when fixing this for that version.

Reply:
Oh wait yeah you're right, if you look at this part of his gen2 PR he removed the analogous terminal transaction code:
https://github.com/okta/okta-signin-widget/pull/3063/files#diff-9d2ede01662be7d3291032cd313089a2f2a7399c19f8f078a7b2910c2a38f141
I have a new Bacon run going right now without the gen3 cookie terminal transaction code so hopefully that passes 🙏

Reply:
Thanks @trevoring-okta for the context -- that's clear to me.